protocol/protocol.go

```diff
@@ -71,3 +71,86 @@ type Responder interface {
 	// The EnvelopeType should be either wire.Reply or wire.Exception.
 	EncodeResponse(v wire.Value, t wire.EnvelopeType, w io.Writer) error
 }
+
+// StreamProtocol defines a specific way for a Thrift value to be
+// encoded or decoded, implemented in a streaming fashion.
+type StreamProtocol interface {
+	StreamWriter(io.Writer) Writer
+	StreamReader(io.Reader) Reader
+}
+
+// FieldHeader defines the metadata needed to define the
+// beginning of a field in a Thrift value.
+type FieldHeader struct {
+	ID int16
+	Type wire.Type
+}
+
+// MapHeader defines the metadata needed to define the
+// beginning of a map in a Thrift value.
+type MapHeader struct {
+	KeyType wire.Type
+	ValueType wire.Type
+	Length int
+}
+
+// SetHeader defines the metadata needed to define the
+// beginning of a set in a Thrift value.
+type SetHeader struct {
+	Length int
+	Type wire.Type
+}
+
+// ListHeader defines the metadata needed to define the
+// beginning of a list in a Thrift value.
+type ListHeader struct {
+	Length int
+	Type wire.Type
+}
+
+// Writer defines an encoder for a Thrift value,
+// implemented in a streaming fashion.
+type Writer interface {
+	WriteBool(bool) error
+	WriteInt8(int8) error
+	WriteInt16(int16) error
+	WriteInt32(int32) error
+	WriteInt64(int64) error
+	WriteString(string) error
+	WriteDouble(float64) error
+	WriteBinary([]byte) error
+	WriteStructBegin() error
+	WriteStructEnd() error
+	WriteFieldBegin(FieldHeader) error
+	WriteFieldEnd() error
+	WriteMapBegin(MapHeader) error
+	WriteMapEnd() error
+	WriteSetBegin(SetHeader) error
+	WriteSetEnd() error
+	WriteListBegin(ListHeader) error
+	WriteListEnd() error
+}
+
+// Reader defines an decoder for a Thrift value,
+// implemented in a streaming fashion.
+type Reader interface {
+	ReadBool() (bool, error)
+	ReadInt8() (int8, error)
+	ReadInt16() (int16, error)
+	ReadInt32() (int32, error)
+	ReadInt64() (int64, error)
+	ReadString() (string, error)
+	ReadDouble() (float64, error)
+	ReadBinary() ([]byte, error)
+	ReadStructBegin() error
+	ReadStructEnd() error
+	ReadFieldBegin() (FieldHeader, bool, error)
+	ReadFieldEnd() error
+	ReadListBegin() (ListHeader, error)
+	ReadListEnd() error
+	ReadSetBegin() (SetHeader, error)
+	ReadSetEnd() error
+	ReadMapBegin() (MapHeader, error)
+	ReadMapEnd() error
+	Skip(wire.Type) error
```

Comment:
I'm not sure about commenting styles for repeated, patterned methods like above, but `Skip` probably warrants its own comment.

Reply:
The repeated methods also seemed somewhat self-explanatory, and it felt like any comments that delved into implementation details felt too much into the weeds, though I'm happy to put in comments if that is the convention.

I did add a small comment for skip!